comms/dht/src/dedup/dedup_cache.rs

```diff
@@ -0,0 +1,185 @@
+                debug!(
+                    target: LOG_TARGET,
+                    "Message dedup cache: msg_count {}, capacity {}, num_removed {}", msg_count, capacity, num_removed,
+                );
+                Ok(num_removed)
+            })
+            .await
+    }
+
+    async fn insert_body_hash(&self, body_hash: String) -> Result<usize, StorageError> {
+        self.connection
+            .with_connection_async(move |conn| {
+                let insert_result = diesel::insert_into(dedup_cache::table)
+                    .values(UpdateDedupCacheSql {
+                        body_hash: Some(body_hash.clone()),
+                        number_of_hits: Some(1),
+                        last_hit_at: Some(DedupCacheDatabase::formatted_naive_date_time()?),
+                    })
+                    .execute(conn)
+                    .map_err(Into::into);
+                match insert_result {
+                    Ok(val) => Ok(val),
+                    Err(diesel::result::Error::DatabaseError(kind, e_info)) => match kind {
+                        DatabaseErrorKind::UniqueViolation => Err(StorageError::UniqueViolation(body_hash)),
+                        _ => Err(diesel::result::Error::DatabaseError(kind, e_info).into()),
+                    },
+                    Err(e) => Err(e.into()),
+                }
+            })
+            .await
+    }
+
+    async fn update_number_of_hits(&self, body_hash: String) -> Result<usize, StorageError> {
+        let record_to_update = self.find_by_body_hash(body_hash.clone()).await?;
+        self.connection
+            .with_connection_async(move |conn| {
+                diesel::update(dedup_cache::table.filter(dedup_cache::body_hash.eq(&body_hash)))
+                    .set(UpdateDedupCacheSql {
+                        body_hash: None,
+                        number_of_hits: Some(record_to_update.number_of_hits + 1),
+                        last_hit_at: Some(DedupCacheDatabase::formatted_naive_date_time()?),
+                    })
+                    .execute(conn)
+                    .map_err(Into::into)
+            })
+            .await
+    }
+
+    async fn find_by_body_hash(&self, body_hash: String) -> Result<DedupCacheSql, StorageError> {
+        self.connection
+            .with_connection_async(move |conn| {
+                Ok(dedup_cache::table
+                    .filter(dedup_cache::body_hash.eq(body_hash))
+                    .first::<DedupCacheSql>(conn)?)
+            })
+            .await
+    }
+
+    fn body_hash_string(body_hash: Vec<u8>) -> String {
+        hex::to_hex(&Challenge::new().chain(body_hash).finalize())
+    }
+
+    fn formatted_naive_date_time() -> Result<NaiveDateTime, StorageError> {
+        NaiveDateTime::parse_from_str(
+            Utc::now().naive_utc().format("%Y-%m-%d %H:%M:%S").to_string().as_str(),
+            "%Y-%m-%d %H:%M:%S",
+        )
+        .map_err(|e| StorageError::ParseError(e.to_string()))
+    }
```

Comment:
?
```suggestion
   Utc::now().naive_utc()
```

Reply:
This function is to make it easier for humans to read the database entries; database field `stored_at` has a default `DEFAULT CURRENT_TIMESTAMP` and its format is automatically entered as `2021-07-23 04:01:14`. When the `last_hit_at` field is populated with `Utc::now().naive_utc()` it prints something like `2021-07-23 04:01:14.235873992`, which makes it difficult to compare visually.

Reply:
I don't buy this reasoning. I would just store the data as is. SQL has the capability to display the data differently via SELECT statements for the poor human users who are struggling

Reply:
`SELECT  STRFTIME('%Y-%m-%d %H:%M:%S"', last_hit_at) as last_hit_ts FROM dedup_cache ORDER BY last_hit_ts` would do it.

Reply:
Happy to merge once this change is applied

Reply:
Happy to remove `fn  formatted_naive_date_time `